File: src/com/haskforce/haskell/project/externalSystem/stack/StackSettings.scala
package com.haskforce.haskell.project.externalSystem.stack

import java.util

import com.intellij.openapi.Disposable
import com.intellij.openapi.components.{PersistentStateComponent, State, Storage}
import com.intellij.openapi.externalSystem.settings.{AbstractExternalSystemSettings, ExternalSystemSettingsListener}
import com.intellij.openapi.project.Project
import com.intellij.util.xmlb.annotations.XCollection

@State(name = "StackSettings", storages = Array(new Storage("haskell-stack.xml")))
final class StackSettings(
  project: Project
) extends AbstractExternalSystemSettings[
    StackSettings,
    StackProjectSettings,
    StackProjectSettingsListener
  ](StackTopic, project)
  with PersistentStateComponent[StackSettings.State] {

  override def getState: StackSettings.State = {
    val state = new StackSettings.State
    fillState(state)
    state
  }

  override def loadState(state: StackSettings.State): Unit = {
    super[AbstractExternalSystemSettings].loadState(state)
  }

  override def subscribe(
    listener: ExternalSystemSettingsListener[StackProjectSettings]
  ): Unit = {
    doSubscribe(
      new StackDelegatingExternalSystemSettingsListener(listener),
      project
    )
  }

  override def subscribe(
    listener: ExternalSystemSettingsListener[StackProjectSettings],
    parentDisposable: Disposable
  ): Unit = {
    doSubscribe(
      new StackDelegatingExternalSystemSettingsListener(listener),
      parentDisposable
    )
  }

  override def copyExtraSettingsFrom(settings: StackSettings): Unit = {}

  override def checkSettings(
    old: StackProjectSettings,
    current: StackProjectSettings
  ): Unit = {
    if (old != current) getPublisher.onStackProjectSettingsChange()
  }
}

object StackSettings {

  def getInstance(project: Project): StackSettings = {
    project.getService(classOf[StackSettings])
  }

  class State extends AbstractExternalSystemSettings.State[StackProjectSettings] {

    private val linkedProjectSettings: util.TreeSet[StackProjectSettings] = {
      new util.TreeSet[StackProjectSettings]()
    }

    @XCollection(elementTypes = Array(classOf[StackProjectSettings]))
    override def getLinkedExternalProjectsSettings: util.Set[StackProjectSettings] = {
      linkedProjectSettings
    }

    override def setLinkedExternalProjectsSettings(settings: util.Set[StackProjectSettings]): Unit = {
      if (settings == null) return
      linkedProjectSettings.addAll(settings)
      ()
    }
  }
}

File: src/com/haskforce/haskell/project/externalSystem/stack/StackSettings.java
package com.haskforce.haskell.project.externalSystem.stack;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.externalSystem.settings.AbstractExternalSystemSettings;
import com.intellij.openapi.externalSystem.settings.ExternalSystemSettingsListener;
import com.intellij.openapi.project.Project;
import com.intellij.util.xmlb.annotations.XCollection;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Set;
import java.util.TreeSet;

@State(name = "StackSettings", storages = @Storage("haskell-stack.xml"))
public class StackSettings
  extends
    AbstractExternalSystemSettings<
      StackSettings,
      StackProjectSettings,
      StackProjectSettingsListener
    >
  implements PersistentStateComponent<StackSettings.State> {

  public StackSettings(@NotNull Project project) {
    super(StackTopic.get(), project);
  }

  @NotNull
  public static StackSettings getInstance(@NotNull Project project) {
    return project.getService(StackSettings.class);
  }

  @Override
  public void initializeComponent() {
    System.out.println("StackSettings.initializeComponent");
  }

  @Override
  public void subscribe(@NotNull ExternalSystemSettingsListener<StackProjectSettings> listener) {
    doSubscribe(
      new StackDelegatingExternalSystemSettingsListener(listener),
      getProject()
    );
  }

  @Override
  public void subscribe(@NotNull ExternalSystemSettingsListener<StackProjectSettings> listener, @NotNull Disposable parentDisposable) {
    doSubscribe(
      new StackDelegatingExternalSystemSettingsListener(listener),
      parentDisposable
    );
  }

  @Override
  protected void copyExtraSettingsFrom(@NotNull StackSettings settings) {
  }

  @Nullable
  @Override
  public State getState() {
    State state = new State();
    fillState(state);
    return state;
  }

  @Override
  public void loadState(@NotNull State state) {
    super.loadState(state);
  }

  @Override
  protected void checkSettings(@NotNull StackProjectSettings old, @NotNull StackProjectSettings current) {
    if (!old.equals(current)) getPublisher().onStackProjectSettingsChange();
  }

  public static class State
    implements AbstractExternalSystemSettings.State<StackProjectSettings> {

    private final Set<StackProjectSettings> myProjectSettings = new TreeSet<>();

    @Override
    @XCollection(elementTypes = {StackProjectSettings.class})
    public Set<StackProjectSettings> getLinkedExternalProjectsSettings() {
      return myProjectSettings;
    }

    @Override
    public void setLinkedExternalProjectsSettings(
      Set<StackProjectSettings> settings
    ) {
      if (settings == null) return;
      myProjectSettings.addAll(settings);
    }
  }
}
